crypto/dilithium/pqcrystals_dilithium_ref_common/sign.c

```diff
@@ -63,73 +68,92 @@ int crypto_sign_keypair(ml_dsa_params *params, uint8_t *pk, uint8_t *sk) {
+*
+* Returns 0 (success) -1 on failure
+**************************************************/
+int crypto_sign_keypair(ml_dsa_params *params, uint8_t *pk, uint8_t *sk) {
+  uint8_t seed[SEEDBYTES];
+  /* Get randomness for rho, rhoprime and key */
+  if (!RAND_bytes(seed, SEEDBYTES)) {
+    return -1;
+  }
+  crypto_sign_keypair_internal(params, pk, sk, seed);
+  return 0;
+}
+
+/*************************************************
+* Name:        crypto_sign_signature_internal
+*
+* Description: FIPS 204: Algorithm 7 ML-DSA.Sign_internal.
+*              Computes signature. Internal API.
+*
+* Arguments:   - ml_dsa_params:  parameter struct
 *              - uint8_t *sig:   pointer to output signature (of length CRYPTO_BYTES)
 *              - size_t *siglen: pointer to output length of signature
 *              - uint8_t *m:     pointer to message to be signed
 *              - size_t mlen:    length of message
-*              - uint8_t *ctx:   pointer to context string
-*              - size_t ctxlen:  length of context string
+*              - uint8_t *pre:   pointer to prefix string
+*              - size_t prelen:  length of prefix string
+*              - uint8_t *rnd:   pointer to random seed
 *              - uint8_t *sk:    pointer to bit-packed secret key
 *
 * Returns 0 (success) or -1 (context string too long)
 **************************************************/
-int crypto_sign_signature(ml_dsa_params *params,
-                          uint8_t *sig,
-                          size_t *siglen,
-                          const uint8_t *m,
-                          size_t mlen,
-                          const uint8_t *ctx,
-                          size_t ctxlen,
-                          const uint8_t *sk)
+int crypto_sign_signature_internal(ml_dsa_params *params,
+                                   uint8_t *sig,
+                                   size_t *siglen,
+                                   const uint8_t *m,
+                                   size_t mlen,
+                                   const uint8_t *pre,
+                                   size_t prelen,
+                                   const uint8_t *rnd,
+                                   const uint8_t *sk)
 {
   unsigned int n;
-  uint8_t seedbuf[2*SEEDBYTES + TRBYTES + RNDBYTES + 2*CRHBYTES];
-  uint8_t *rho, *tr, *key, *mu, *rhoprime, *rnd;
+  uint8_t seedbuf[2*SEEDBYTES + TRBYTES + 2*CRHBYTES];
+  uint8_t *rho, *tr, *key, *mu, *rhoprime;
   uint16_t nonce = 0;
   polyvecl mat[DILITHIUM_K_MAX], s1, y, z;
   polyveck t0, s2, w1, w0, h;
   poly cp;
   keccak_state state;
 
-  if(ctxlen > 255)
-    return -1;
-
   rho = seedbuf;
   tr = rho + SEEDBYTES;
   key = tr + TRBYTES;
-  rnd = key + SEEDBYTES;
-  mu = rnd + RNDBYTES;
+  mu = key + SEEDBYTES;
   rhoprime = mu + CRHBYTES;
   unpack_sk(params, rho, tr, key, &t0, &s1, &s2, sk);
 
-  /* Compute mu = CRH(tr, 0, ctxlen, ctx, msg) */
-  mu[0] = 0;
-  mu[1] = ctxlen;
+  /* Compute mu = CRH(tr, pre, msg) */
```

Comment:
can you add references to lines in FIPS 204, Algorithm 7 please (for verify_internal and keygen_internal as well)? For example, I believe this comment refers to line 6 that says: 𝜇 ← H(BytesToBits(𝑡𝑟)||𝑀 ′, 64) ?

Reply:
Yes, I'm staging a PR that adds a LOT of documentation and call out to each of the algorithms in FIPS 204 and their implementation here. 

However, I want to get the larger code integrations complete before I start touching every single file in the dilithium repo to add comments. This will prevent me from re-base/conflicts pain down the line.

I'll add specific call outs to the lines in sign/verify in that forthcoming PR.

Reply:
sure, but the point of documentation is to make the review easier, once we merge the code it's not as useful :)

Reply:
Ahh I understand now, added these callouts to lines in the spec in [8c3d53e](https://github.com/aws/aws-lc/pull/1999/commits/8c3d53e8abe8cdd6783c64bc34e98352d44a83fd). I plan on adding a lot more documentation like this.